src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql

```diff
@@ -0,0 +1,441 @@
+    ) BEGIN
+        THROW 50404, 'Task not exist or runid not match', 1;
+    END
+
+    -- We will timestamp the jobs when we update them to track stale jobs.
+    DECLARE @heartbeatDateTime datetime2(7) = SYSUTCDATETIME()
+
+	UPDATE dbo.TaskInfo
+	SET Status = 3, HeartbeatDateTime = @heartbeatDateTime, Result = @taskResult
+	WHERE TaskId = @taskId
+
+    SELECT TaskId, QueueId, Status, TaskTypeId, RunId, IsCanceled, RetryCount, HeartbeatDateTime, InputData, TaskContext, Result
+	FROM [dbo].[TaskInfo]
+	where TaskId = @taskId
+
+    COMMIT TRANSACTION
+GO
+
+
+/*************************************************************
+    Stored procedures for cancel task
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     CancelTask
+--
+-- DESCRIPTION
+--     Cancel the task and update task status.
+--
+-- PARAMETERS
+--     @taskId
+--         * The ID of the task record
+--
+CREATE OR ALTER PROCEDURE [dbo].[CancelTask]
+    @taskId varchar(64)
+AS
+    SET NOCOUNT ON
+
+    SET XACT_ABORT ON
+    BEGIN TRANSACTION
+	
+    -- We will timestamp the jobs when we update them to track stale jobs.
+    DECLARE @heartbeatDateTime datetime2(7) = SYSUTCDATETIME()
+
+    IF NOT EXISTS
+    (
+        SELECT *
+        FROM [dbo].[TaskInfo]
+        WHERE TaskId = @taskId
+    ) BEGIN
+        THROW 50404, 'Task not exist', 1;
+    END
+
+	UPDATE dbo.TaskInfo
+	SET IsCanceled = 1, HeartbeatDateTime = @heartbeatDateTime
+	WHERE TaskId = @taskId
+
+    SELECT TaskId, QueueId, Status, TaskTypeId, RunId, IsCanceled, RetryCount, HeartbeatDateTime, InputData, TaskContext, Result
+	FROM [dbo].[TaskInfo]
+	where TaskId = @taskId
+
+    COMMIT TRANSACTION
+GO
+
+
+/*************************************************************
+    Stored procedures for reset task
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     ResetTask
+--
+-- DESCRIPTION
+--     Reset the task status.
+--
+-- PARAMETERS
+--     @taskId
+--         * The ID of the task record
+--     @runId
+--         * Current runId for this exuction of the task
+--
+CREATE OR ALTER PROCEDURE [dbo].[ResetTask]
+    @taskId varchar(64),
+    @runId varchar(50),
+    @result varchar(max),
+    @maxRetryCount smallint = 3
+AS
+    SET NOCOUNT ON
+
+    SET XACT_ABORT ON
+    BEGIN TRANSACTION
+	
+    -- Can only reset task with same runid
+    DECLARE @retryCount smallint
+    DECLARE @status smallint
+
+    SELECT @retryCount = RetryCount, @status = Status
+    FROM [dbo].[TaskInfo]
+    WHERE TaskId = @taskId and RunId = @runId
+
+	-- We will timestamp the jobs when we update them to track stale jobs.
+    DECLARE @heartbeatDateTime datetime2(7) = SYSUTCDATETIME()
+
+    IF (@retryCount IS NULL) BEGIN
+        THROW 50404, 'Task not exist or runid not match', 1;
+    END
+
+    IF (@retryCount >= @maxRetryCount) BEGIN
+		UPDATE dbo.TaskInfo
+		SET Status = 3, HeartbeatDateTime = @heartbeatDateTime, Result = @result
```

Comment:
>Result = @result [](start = 58, length = 16)

Should we update the `Result` here even though we are beyond `maxRetryCount`?

Reply:
Update result here is for the store the failed task result. Client would get failed result details for the last retry.